File: app/typescript/src/domain/purchase/unit-price.repository.ts
import { PrismaClient } from '@prisma/client'

// 単価マスタの入力型
export interface CreateUnitPriceInput {
  itemCode: string
  supplierCode: string
  lotUnitQuantity?: number
  effectiveFrom: Date
  effectiveTo?: Date
  unitPrice: number
  createdBy?: string
}

// 単価マスタの出力型
export interface UnitPrice {
  id: number
  itemCode: string
  supplierCode: string
  lotUnitQuantity: number
  effectiveFrom: Date
  effectiveTo: Date | null
  unitPrice: number
  createdAt: Date
  createdBy: string | null
  updatedAt: Date
  updatedBy: string | null
}

export class UnitPriceRepository {
  constructor(private prisma: PrismaClient) {}

  async create(input: CreateUnitPriceInput): Promise<UnitPrice> {
    const result = await this.prisma.$queryRaw<UnitPrice[]>`
      INSERT INTO unit_prices (
        item_code, supplier_code, lot_unit_quantity, effective_from,
        effective_to, unit_price, created_by
      )
      VALUES (
        ${input.itemCode}, ${input.supplierCode}, ${input.lotUnitQuantity ?? 1},
        ${input.effectiveFrom}, ${input.effectiveTo ?? null},
        ${input.unitPrice}, ${input.createdBy ?? null}
      )
      RETURNING
        id,
        item_code as "itemCode",
        supplier_code as "supplierCode",
        lot_unit_quantity as "lotUnitQuantity",
        effective_from as "effectiveFrom",
        effective_to as "effectiveTo",
        unit_price as "unitPrice",
        created_at as "createdAt",
        created_by as "createdBy",
        updated_at as "updatedAt",
        updated_by as "updatedBy"
    `
    return result[0]
  }

  async findCurrentPrice(
    itemCode: string,
    supplierCode: string,
    lotUnitQuantity: number,
    targetDate: Date
  ): Promise<UnitPrice | null> {
    const result = await this.prisma.$queryRaw<UnitPrice[]>`
      SELECT
        id,
        item_code as "itemCode",
        supplier_code as "supplierCode",
        lot_unit_quantity as "lotUnitQuantity",
        effective_from as "effectiveFrom",
        effective_to as "effectiveTo",
        unit_price as "unitPrice",
        created_at as "createdAt",
        created_by as "createdBy",
        updated_at as "updatedAt",
        updated_by as "updatedBy"
      FROM unit_prices
      WHERE item_code = ${itemCode}
        AND supplier_code = ${supplierCode}
        AND lot_unit_quantity = ${lotUnitQuantity}
        AND effective_from <= ${targetDate}
        AND (effective_to IS NULL OR effective_to >= ${targetDate})
      ORDER BY effective_from DESC
      LIMIT 1
    `
    return result.length > 0 ? result[0] : null
  }

  async findByItemAndSupplier(
    itemCode: string,
    supplierCode: string
  ): Promise<UnitPrice[]> {
    const result = await this.prisma.$queryRaw<UnitPrice[]>`
      SELECT
        id,
        item_code as "itemCode",
        supplier_code as "supplierCode",
        lot_unit_quantity as "lotUnitQuantity",
        effective_from as "effectiveFrom",
        effective_to as "effectiveTo",
        unit_price as "unitPrice",
        created_at as "createdAt",
        created_by as "createdBy",
        updated_at as "updatedAt",
        updated_by as "updatedBy"
      FROM unit_prices
      WHERE item_code = ${itemCode}
        AND supplier_code = ${supplierCode}
      ORDER BY lot_unit_quantity, effective_from DESC
    `
    return result
  }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Persistence/Repositories/UnitPriceRepository.cs
using Dapper;
using Npgsql;
using ProductionManagement.Application.Port.Out;
using ProductionManagement.Domain.Models.Purchase;
using ProductionManagement.Infrastructure.Persistence.TypeHandlers;

namespace ProductionManagement.Infrastructure.Persistence.Repositories;

/// <summary>
/// 単価マスタリポジトリ実装
/// </summary>
public class UnitPriceRepository : IUnitPriceRepository
{
    private readonly string _connectionString;

    static UnitPriceRepository()
    {
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
    }

    public UnitPriceRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task SaveAsync(UnitPrice unitPrice)
    {
        const string sql = """
            INSERT INTO "単価マスタ" (
                "品目コード", "取引先コード", "ロット単位数", "使用開始日",
                "使用停止日", "単価", "作成者"
            ) VALUES (
                @ItemCode, @SupplierCode, @LotUnitQuantity, @EffectiveFrom,
                @EffectiveTo, @Price, @CreatedBy
            )
            RETURNING "ID"
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        unitPrice.Id = await connection.ExecuteScalarAsync<int>(sql, unitPrice);
    }

    public async Task<UnitPrice?> FindEffectiveUnitPriceAsync(
        string itemCode, string supplierCode, DateOnly date)
    {
        const string sql = """
            SELECT
                "ID" AS Id,
                "品目コード" AS ItemCode,
                "取引先コード" AS SupplierCode,
                "ロット単位数" AS LotUnitQuantity,
                "使用開始日" AS EffectiveFrom,
                "使用停止日" AS EffectiveTo,
                "単価" AS Price,
                "作成日時" AS CreatedAt,
                "作成者" AS CreatedBy,
                "更新日時" AS UpdatedAt,
                "更新者" AS UpdatedBy
            FROM "単価マスタ"
            WHERE "品目コード" = @ItemCode
              AND "取引先コード" = @SupplierCode
              AND "使用開始日" <= @Date
              AND ("使用停止日" IS NULL OR "使用停止日" >= @Date)
            ORDER BY "使用開始日" DESC
            LIMIT 1
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        return await connection.QueryFirstOrDefaultAsync<UnitPrice>(
            sql, new { ItemCode = itemCode, SupplierCode = supplierCode, Date = date });
    }

    public async Task DeleteAllAsync()
    {
        const string sql = """DELETE FROM "単価マスタ" """;

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql);
    }
}
